File: compositePattern/csharp/CompositeEquipment.cs
abstract class CompositeEquipment : Equipment
{
	protected List<Equipment> _equipments = new();

	protected CompositeEquipment()
	{
		_name = "CompositeEquipment";
	}

	public override void Add(Equipment equipment)
	{
		_equipments.Add(equipment);
	}

	public override void Remove(Equipment equipment)
	{
		_equipments.Remove(equipment);
	}

	public override int NetPrice()
	{
		var totalNetPrice = 0;

		// Can use iterator pattern here
		foreach (var equipment in _equipments)
		{
			totalNetPrice += equipment.NetPrice();
		}

		return totalNetPrice;
	}

	public override int Power()
	{
		var totalPowerInWatts = 0;

		// Can use iterator pattern here
		foreach (var equipment in _equipments)
		{
			totalPowerInWatts += equipment.Power();
		}

		return totalPowerInWatts;
	}
}

// 機殼
class Chassis : CompositeEquipment
{
	public Chassis()
	{
		_name = "Chassis";
	}

	public override void Add(Equipment equipment)
	{
		List<String> acceptableEquipments = new List<string> { "Motherboard", "Cpu" };
		if (acceptableEquipments.Contains(equipment.GetName()))
		{
			_equipments.Add(equipment);
			return;
		}

		System.Console.WriteLine("Our Chassis can not contain useless equipment!!");
	}
}

File: compositePattern/compositeEquipment.cpp
#include <iostream>
#include "./compositeEquipment.h"

CompositeEquipment::CompositeEquipment()
{
	_name = "CompositeEquipment";
}

void CompositeEquipment::Add(Equipment *equipment)
{
	_equipments.push_back(equipment);
}

void CompositeEquipment::Remove(Equipment *equipment)
{
	_equipments.remove(equipment);
}

int CompositeEquipment::NetPrice()
{
	int totalNetPrice = 0;

	std::list<Equipment *>::iterator iterator = _equipments.begin();
	while (iterator != _equipments.end())
	{
		totalNetPrice += (*iterator)->NetPrice();
		++iterator;
	}

	delete *iterator;
	return totalNetPrice;
}

int CompositeEquipment::Power()
{
	int totalPowerInWatts = 0;

	std::list<Equipment *>::iterator iterator = _equipments.begin();
	while (iterator != _equipments.end())
	{
		totalPowerInWatts += (*iterator)->Power();
		++iterator;
	}

	delete *iterator;
	return totalPowerInWatts;
}

Chassis::Chassis()
{
	_name = "Chassis";
}

void Chassis::Add(Equipment *equipment)
{
	bool isAcceptable = false;
	std::string acceptableEquipments[2] = {"Motherboard", "Cpu"};
	for (std::string ae : acceptableEquipments)
	{
		if (equipment->GetName() == ae)
		{
			_equipments.push_back(equipment);
			isAcceptable = true;
		}
	}

	if (!isAcceptable)
	{
		std::cout << "Our Chassis can not contain useless equipment!!" << std::endl;
	}
}